### PR TITLE
[BUGFIX] fix unknown parameter shapes when np_shape is turned on.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
When np_shape is turned on, the unknown parameter shapes should be -1 instead of 0. So when a parameter is created but isn't initialized, the dimension of unknown size is -1 when np_shape is turned on, and is 0, otherwise.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: python/mxnet/gluon/parameter.py [hunk 1]
```diff
@@ -156,7 +157,10 @@ def grad_req(self, req):
 
     @property
     def shape(self):
-        return self._shape
+        if is_np_shape():
+            return tuple(i if i != 0 else -1 for i in self._shape)
```
Comment (reminisce) [inline python/mxnet/gluon/parameter.py]: I think you also need to handle `self._shape is None`.

File: tests/python/unittest/test_gluon.py [hunk 2]
```diff
@@ -2726,6 +2726,23 @@ def hybrid_forward(self, F, x):
...
+@with_seed()
+def test_np_shape_parameters():
+    class Foo(gluon.Block):
+        def __init__(self, **kwargs):
+            super(Foo, self).__init__(**kwargs)
+            self.dense = gluon.nn.Dense(16)
+        def forward(self, x):
+            return self.dense(x)
+
+    mx.set_np_shape(True)
```
Comment (reminisce) [inline tests/python/unittest/test_gluon.py]: It would be safer to use:
```python
with mx.np_shape(True):
    # test code
```

**Comment by reminisce** [comment]
I think the following two places checking whether the shape is known should be modified too.
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/parameter.py#L272
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/parameter.py#L383

**Comment by reminisce** [comment]
I'm worried about this line
https://github.com/apache/incubator-mxnet/blob/5fc4fc53df74f276aafa51208142e657e9cfe42d/python/mxnet/gluon/parameter.py#L168

Users may notice that we started to use `-1` to represent unknown dim size in np shape semantics and some of them in practice may use shapes like `(-1, 10)` to create parameters for deferred initialization, instead of `(0, 10)`. In those cases, we need to support that unknown dim size can be either `0` or `-1`, so this line should become
```python
all(j in (-1, 0, i) for i, j in zip(new_shape, self._shape))
```
when `is_np_shape = True`.

@szha What do you think?

**Comment by abhinavs95** [comment]
@mxnet-label-bot add [Gluon]

**Comment by szha** [comment]
@reminisce I had similar concern though I thought the API doc should do the job. another option is to expose an internal interface for infer_shape to use, instead of piggybacking on the `.shape` attribute

**Comment by roywei** [comment]
@reminisce @sza @zheng-da  what's the next step here? There is nothing to prevent users from using `-1` as unknown dim. so maybe state clearly in comment or API doc and note experimental?

**Comment by reminisce** [comment]
I think it would be safer to note `np_shape` as experimental and still under development.

**Comment by zheng-da** [comment]
where should we document `-1` is unknown dim and is experimental?

File: python/mxnet/gluon/parameter.py [hunk 3]
```diff
@@ -156,7 +157,12 @@ def grad_req(self, req):
 
     @property
     def shape(self):
-        return self._shape
+        if self._shape is None:
+            return None
+        elif is_np_shape():
+            return tuple(i if i != 0 else -1 for i in self._shape)
```
Comment (roywei) [inline python/mxnet/gluon/parameter.py]: a simple comment about -1

File: python/mxnet/gluon/utils.py [hunk 4]
```diff
@@ -412,3 +413,17 @@ def __enter__(self):
 
     def __exit__(self, ptype, value, trace):
         self.detach()
+
+def shape_is_known(shape):
+    """Check whether a shape is completely known w/ or w/o np semantics."""
```
Comment (roywei) [inline python/mxnet/gluon/utils.py]: maybe a reference to np shape for more details.
